### PR TITLE
Split GI effects and fix stereoscopic rendering of GI effects

### PR DESCRIPTION
**Pull request by BastiaanOlij**
This PR does a couple of things:
- Split out Voxel GI storage
- Split out effects used by GI into a now resolve class
- Adds stereoscopic support to Voxel GI and SDFGI (note that many other effects haven't had this treatment yet so only test this in isolation! I will get to the rest soon)

I have also decided to make `RendererSceneGIRD` a subclass of the new `RendererVoxelGIStorage` class so storage and implementation is now a single implementation, this is something that Reduz hinted at (if I understood him correctly) and I feel this change makes sense to keep things together. I've also namespaced and renamed the class to `RendererRD::GRID`.
The file name is still the same, I thought about renaming it but I'm worried about rebasing issues. Might still rename 'renderer_scene_gi_rd.h/cpp`

### CODE REVIEW

**Comment by BastiaanOlij** [comment]
Only just beginning with this. Started with getting the debug functions to work so can start diagnosing where things are not working properly. SDF probes and cascades are rendered in world space so these should not be effected by stereoscopic rendering but something is definately broken:

Probes in normal window:
image

Probes in multiview (using Mobile VR Interface with k1 and k2 set to 0, so no lens distortion):
image

**Comment by BastiaanOlij** [comment]
Cascades in editor:
image

And in stereo:
image

Definitely not working as it should :P I think the weird proportion issue is because we've not actually passing the projection matrix, we're just passing a few elements which I think will only be correct for symmetric projection matrices, not for asymmetric ones.

**Comment by BastiaanOlij** [comment]
Ok, properly un-projecting now in the debug view... seeing its debug view I'm guessing the slight overhead to the simplified option is acceptable.

**Comment by BastiaanOlij** [comment]
Ok, so depth pass skips the stuff it needs to do.. so that is my next step, make multiview versions of `PASS_MODE_DEPTH_NORMAL_ROUGHNESS` and `PASS_MODE_DEPTH_NORMAL_ROUGHNESS_VOXEL_GI`

**Comment by BastiaanOlij** [comment]
Got SDFGI pretty much working:

image

Next up is:
- looking at voxel Gi and making it work in stereo
- splitting out the related effects more properly
- possibly looking at splitting out storage or making storage and gi classes work better together
- bunch of cleanup, revisiting this code has made me realise a lot of the split logic I did for GI a year or two ago I overlooked a few things

**Comment by lyuma** [comment]
Hi Mux,
I've been slowly learning the various renderer code over the past few days to give this a review and isolate the places where it is failing. I'm still working at it, so this is a rather incomplete review, but I thought I'd give some of my feedback thus far.

I made a change on top of this patch to try and solve one problem: I implemented Compute Shader (copy.glsl) support for copying array textures, used for non-MSAA depth texture: https://github.com/lyuma/godot/commit/622b7d42b9343bd9de488105102dafc5b13fbaef (it's based on a bunch of other changes, so it needs to be cherry-picked):
I was trying to fix an issue with shadows unrelated to GI... but I think some of the GI debug textures use this function as well, so it might also help get those working.

In terms of code style, the refactor made it a bit harder for me to understand and review the change because some code was moved around, but I'm starting to find my way. I think the class named "GRID" probably should be named "GIRD" (as in global illumination rendering device). "GIRD" shows up in the code in a few places.

In terms of the code itself, I am worried about the amount of things that are being rendered per-eye in stereo. The GI should not be calculated in each eye. SDFGI does depend on the camera position, yes, but it only depends on it to provide a series of cascades. The performance of SDFGI is already too heavy, so if it requires rendering per-eye, it is not a good idea in this form. In my opinion, the two eyes should be close enough that they can share cascades.

This is why I am really worried about some of the changes in gi_rd... (I'm not talking about the debug views, since they are just for debugging.)

This patch seems to reduce the difference in both eyes for SDFGI, so in that sense it's a step in the right direction. However, it seems to introduce a few *new* big issues which were not present before applying this PR:

1. SDFGI is now wrong in both the left eye and the right eye, though it's wrong in a different way than before. My suspicion is something was changed to use multiview that should not be per-eye.

Image showing horizontal lines in GI for both eyes

2. MSAA causes the editor to hang with "Did not create swapchain successfully" printed to console repeatedly
```
ERROR: Attempted to free invalid ID: 651688272733083
   at: RenderingDeviceVulkan::_free_internal (drivers\vulkan\rendering_device_vulkan.cpp:8572)
Allocating GI buffers
ERROR: Vulkan: Did not create swapchain successfully.
   at: (drivers\vulkan\vulkan_context.cpp:1959)
ERROR: Condition "err" is true. Returning: ERR_CANT_CREATE
   at: VulkanContext::swap_buffers (drivers\vulkan\vulkan_context.cpp:2036)
```

3. `Invalid ambient buffer` is spammed in the editor every other time I check and uncheck VoxelGI or SDFGI. It's not consistent. sometimes it goes away if I toggle on a GI setting in WorldEnvironment.

Thanks for all your work.

**Comment by BastiaanOlij** [comment]
Nice catching the rename issue, yeah it should have been `GIRD`. Seeing we're moving things into the `RendererRD` namespace it doesn't make sense to have such convoluted names.

The "Invalid ambient buffer" was just me testing, that line of code should be removed. There are valid conditions where there isn't an ambient buffer yet, but I also had some situations where it wasn't triggering rendering the ambient and reflection buffers.

I'll do some testing on MSAA, I may have overlooked something there.

As for SDFGI , I need to have a closer look but I think the cascades are rendered in world space, not for each eye. There is then a render pass that generates screen space buffers that I don't think we can get away with doing in anything but stereo.

**Comment by BastiaanOlij** [comment]
Only had a quick look earlier today, with MSAA it is going wrong in the resolve shader. Set 0 has two bindings, but when configuring the uniform it's complaining it's only expecting 1 binding. Looks like something is not being read from the compiled shader properly

File: drivers/vulkan/rendering_device_vulkan.cpp [hunk 1]
```diff
@@ -2314,6 +2314,12 @@ RID RenderingDeviceVulkan::texture_create_shared_from_slice(const TextureView &p
 		image_view_create_info.viewType = VK_IMAGE_VIEW_TYPE_2D_ARRAY;
 	}
 
+	if (p_slice_type == TEXTURE_SLICE_2D) {
+		texture.type = TEXTURE_TYPE_2D;
```
Comment (BastiaanOlij) [inline drivers/vulkan/rendering_device_vulkan.cpp:2318]: @clayjohn @JFonS @reduz I ran into issues with some of the checking in various places that when creating a slice of a `TEXTURE_TYPE_2D_ARRAY`, the texture.type of the slice was still identified as an array, even though it no longer is.

This fixes this and seems to work but wanted to point it out in case you guys in case this has effects I havent thought off.

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp [hunk 2]
```diff
@@ -227,6 +272,22 @@ void RenderForwardClustered::RenderBufferDataForwardClustered::configure(RID p_c
...
+		if (view_count == 1) {
+			// just reuse
+			color_views[0] = color;
+			depth_views[0] = depth;
+			color_msaa_views[0] = color_msaa;
+			depth_msaa_views[0] = depth_msaa;
+		} else {
+			// create slices
+			for (uint32_t v = 0; v < view_count; v++) {
+				color_views[v] = RD::get_singleton()->texture_create_shared_from_slice(RD::TextureView(), color, v, 0);
```
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:284]: I'm recreating slices on the color and depth buffers here because the slices are only available outside of the renderer implementation. This we should clean up once we reorganize the renderer logic so we have access to members of our main render buffer.

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp [hunk 3]
```diff
@@ -1490,17 +1565,18 @@ void RenderForwardClustered::_render_scene(RenderDataRD *p_render_data, const Co
 				if (needs_pre_resolve) {
 					RD::get_singleton()->barrier(RD::BARRIER_MASK_RASTER, RD::BARRIER_MASK_COMPUTE);
 				}
-				storage->get_effects()->resolve_gi(render_buffer->depth_msaa, render_buffer->normal_roughness_buffer_msaa, using_voxelgi ? render_buffer->voxelgi_buffer_msaa : RID(), render_buffer->depth, render_buffer->normal_roughness_buffer, using_voxelgi ? render_buffer->voxelgi_buffer : RID(), Vector2i(render_buffer->width, render_buffer->height), texture_multisamples[render_buffer->msaa]);
+				resolve_effects->resolve_gi(render_buffer->depth_msaa, render_buffer->normal_roughness_buffer_msaa, using_voxelgi ? render_buffer->voxelgi_buffer_msaa : RID(), render_buffer->depth, render_buffer->normal_roughness_buffer, using_voxelgi ? render_buffer->voxelgi_buffer : RID(), Vector2i(render_buffer->width, render_buffer->height), texture_multisamples[render_buffer->msaa]);
```
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:1568]: This might need stereo support too? I'm not sure..

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp [hunk 4]
```diff
@@ -1593,14 +1672,18 @@ void RenderForwardClustered::_render_scene(RenderDataRD *p_render_data, const Co
 	}
 
 	if (render_buffer && !can_continue_color && render_buffer->msaa != RS::VIEWPORT_MSAA_DISABLED) {
-		RD::get_singleton()->texture_resolve_multisample(render_buffer->color_msaa, render_buffer->color);
+		// Handle views individual, might want to look at rewriting our resolve to do both layers in one pass.
+		for (uint32_t v = 0; v < render_buffer->view_count; v++) {
+			RD::get_singleton()->texture_resolve_multisample(render_buffer->color_msaa_views[v], render_buffer->color_views[v]);
+		}
+		// TODO mame this do multiview
```
Comment (BastiaanOlij) [inline servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:1679]: I've not changed the specular resolve, that may need more work for stereo support and that should be done in a separate PR.

File: servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp (hunk 3 above)
Comment (BastiaanOlij) [reply servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:1568]: ok this needs stereo support, looking into it..

**Comment by BastiaanOlij** [comment]
@lyuma I think I found all the places that were broken with MSAA now so please retest.

I didn't look into improving the compute shaders to just handle multiple layers instead of handling each layer one at a time. I'll have a look at that later in the week but might do that as an improvement in a separate PR.

**Comment by reduz** [comment]
Great work! Although there are some concerns I have:
* All the code (variable, comments, etc) is written as if XR and Multiview were what is used as the default and mono was the exception. We discussed this already and decided this is very confusing. Code should be written explicitly as if Multiview is the exception, so code, variables and comments must reflect this.
* I don't understand why eye offset can't be used.

**Review by clayjohn — COMMENTED** [review]
I thought we settled on the name renderer_gi?

I don't think we should mix GI in with the effects stuff as it is more than just an effect, it includes storage and implementation for a more complex system.

**Comment by BastiaanOlij** [comment]
> I thought we settled on the name renderer_gi?

Seeing we're namespacing the implementation I ran into some naming issues there hence rendering_gi_effect on the base class and just gi_effect on the implementation. But I can change it, I'm not married on it in any way :)

We did discuss moving it into effects, I got the impression we did have consensus on that as it made more sense than having it in storage. Seeing we'll do something similar to fog and may need to think of the same for certain other parts that are still mixed into the bigger classes I am a fan of moving these things into a separate folder even if we don't move it into effects.

**Review by clayjohn — COMMENTED** [review]
Looks good to me!

Before merging let's get @reduz to review the change in RenderingDevice and @lyuma to confirm that the VR/Multiview specific changes are good

**Review by reduz — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
